role-make-version-changelog.sh

```diff
@@ -130,50 +137,108 @@ else
     fi
 fi
 if [ "$skip" = false ]; then
-    echo "If you want to continue, enter the new tag in the form ${allow_v}X.Y.Z"
-    echo "where X, Y, and Z are integers corresponding to the semantic"
-    echo "version based on the changes above."
-    echo "Or, just press Enter to skip this role and go to the next role."
+    if [[ "$latest_tag" =~ ^"$allow_v"([0-9]+)[.]([0-9]+)[.]([0-9]+)$ ]]; then
+        ver_major="${BASH_REMATCH[1]}"
+        ver_minor="${BASH_REMATCH[2]}"
+        ver_patch="${BASH_REMATCH[3]}"
+    elif [ -z "$latest_tag" ]; then
+        ver_major=0
+        ver_minor=0
+        ver_patch=0
+    else
+        echo ERROR: unexpected tag "$latest_tag"
+        exit 1
+    fi
+    if [ -z "$latest_tag" ]; then
+        commit_range=HEAD
+    else
+        commit_range="${latest_tag}.."
+    fi
+    commits=$(git log --pretty=format:%s --no-merges "$commit_range")
+    if echo "$commits" | grep -q '^.*\!:.*'; then
```

Comment:
```suggestion
    breaking_pattern='(^|'$'\n'')[a-z]+!:'
    if [[ "$commits" == $breaking_pattern ]]; then
```
it is possible to avoid pipe and grep, but it is kinda ugly - note that bash pattern matching cannot do multiline strings, so this regex means "pattern beginning with line start OR newline, followed by lower case letters, followed by !:

If this is too ugly for you, keep the grep

Reply:
$commits in this case is commit titles for the $commit_range, I piped to grep here to do a multiline search specifically. I can read $commits line by line with `while read` but that's much more difficult.

Reply:
> I can write `$commits` into a file and grep it instead. Is it better than piping? I guess the issue with the pipe is that it's hard to troubleshoot issues inside of it, right?

Yes, that's one of the reasons.

Reply:
Using a tempfile would make it easier to troubleshoot too

Reply:
Also - if you create a file, then maybe you could also use it below instead of generating the list again in `reverse_commits=$(git log --no-merges --reverse --pretty=format:"%h %s" "$commit_range")`